0000-ilp-over-http.md

```diff
@@ -0,0 +1,59 @@
+---
+title: ILP Over HTTP
+draft: 1
+---
+
+# ILP Over HTTP
+> A bilateral communication protocol for server-to-server connections
+
+## Motivation
+
+Scaling Interledger infrastructure to handle large volumes of ILP packets requires horizontally scaling connectors. Using HTTP for bilateral communication enables service providers to leverage standard tools and services for hosting, load balancing, Distributed Denial of Service (DDoS) protection, and monitoring.
+
+## Overview
+
+In an ILP Over HTTP connection, both peers run HTTP servers with accessible HTTPS endpoints. When peering, the peers exchange their respective URLs, authentication tokens or TLS certificates, ILP addresses, and settlement-related details.
+
+Each ILP Prepare packet is sent as the body of an HTTP request to the peer's server endpoint. ILP Fulfill or Reject packets are returned as the body of the HTTP response.
+
+## Specification
+
+This is a minimal protocol built on HTTP. HTTP/2 is HIGHLY RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Authentication
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests. Bearer Tokens or TLS Client Certificates are RECOMMENDED.
+
+### Request
+
+```http
+POST /ilp HTTP/x.x
```

Comment:
I'm still not convinced about this format vs simply putting the packet headers as HTTP headers and the `data` payload as the data.

The currently method has 2 major disadvantages:
  1. Lots of redundancy if you want to put the ILP address in a header and the packet (this is often the bulk of the packet size already)
  2. Double serialization. You are serializing the packet as OER and then again into an HTTP message.

@justmoon cited the risk of packet corruption but I'm not sure that risk is realistic. This is intended to only be used over HTTPS so the data can't be changed at all between the endpoints.

I realize there is a risk that SSL terminating proxies or load balancers may add or remove headers but they never change existing headers, especially ones they don't understand.

I think it would be worth testing both formats and validating some of these assumptions but in theory this is going to be a lot less efficient.

Reply:
> Double serialization. You are serializing the packet as OER and then again into an HTTP message.

This change would cause us to deserialize the packet more times, only for it to be reserialized when it's passed to the connector. Unless we changed the ledger plugin interface to pass around the parsed packet (rather than buffers), I don't think this is a good idea.

Reply:
> Unless we changed the ledger plugin interface to pass around the parsed packet (rather than buffers), I don't think this is a good idea.

We've already done this in the connector refactor for this exact reason. The current interface is too opinionated on how the packet will be sent over the wire.

Reply:
I'm also a fan of changing the implementations to pass around a deserialized representation (or making this change in any new implementations if we don't want to change the JS one). We made the interface "send data" to begin with with the idea that you might pass different types of data through that. However, there are so many places that deserialize the packet and only really work if it is an ILP packet that the separation of the layers doesn't really work.

Reply:
Regardless of which way you lean, this is really a larger discussion that probably shouldn't be decided in this PR. If we decide to make that change across the board then it would be big enough to justify a change in this protocol